kernel/src/metaptr.rs

```diff
@@ -0,0 +1,113 @@
+// Licensed under the Apache License, Version 2.0 or the MIT License.
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+// Copyright Google LLC 2024.
+
+//! Defines the MetaPtr type
+
+use core::fmt::{Formatter, LowerHex, UpperHex};
+use core::ops::AddAssign;
+
+/// A pointer with target specific metadata.
+/// This should be used any time the kernel wishes to grant authority to the user, or any time
+/// the user should be required to prove validity of a pointer.
+#[derive(Default, Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Hash, Debug)]
+#[repr(transparent)]
+pub struct MetaPtr {
+    ptr: usize,
+}
+
+#[derive(Copy, Clone, PartialEq)]
+pub enum MetaPermissions {
+    Any,
+    Read,
+    Write,
+    ReadWrite,
+    Execute,
+}
+
+impl From<MetaPtr> for usize {
+    #[inline]
+    fn from(from: MetaPtr) -> Self {
+        from.ptr
+    }
+}
+
+impl From<usize> for MetaPtr {
```

Comment:
Hrm... this seems to break the abstraction. E.g., is this OK under CHERI? Why is it ok to created a `MetaPtr` from any-old `usize`?

Reply:
It's OK assuming the metadata would say something like "invalid" or "just a usize".

On CHERI this would correspond to deriving from the NULL capability.

Reply:
I think those will be the semantics. If somebody wanted to maintain provenance, they should not go through usize.  

`new_with_metadata` is an interesting one. While it seems to have a ptr argument, one might guess that the resulting thing would have its provenance. This is true in this patch.  In the CHERI implementation we do the equivalent of:

```
secret_internal_ptr.with_addr(ptr.addr)
```

This is because it is unlikely that the kernel would actually have provided something of the correct provenance, and that method exists to attach it (using base/length/perms to select an appropriate internal thing to derive from).

Currently nothing in this file mentions provenance, so I wonder if this interaction should be documented at all.

Reply:
[Strict provenance will be stable soon](https://github.com/rust-lang/rust/pull/130350) so we should probably document how these functions affect provenance.

How does `new_with_metadata` compare with [`core::ptr::with_exposed_provenance`](https://doc.rust-lang.org/nightly/core/ptr/fn.with_exposed_provenance.html)? They sound similar, but the latter API has been said to be incompatible with CHERI. Is `new_with_metadata` a hybrid CHERI-only mechanism or is it different?

Reply:
Which provenance the pointer returned from `new_with_metadata` is up to its implementation.

On non-CHERI, I suspect we will return the input.

In the CHERI implementation, we will derive from one of a global DDC, or a global PCC (that covers the entire address space, and all objects are derived from). I think we might even see it in a Purecap world, because we might have a pointer that is RW and need to derive one with the same bounds but with RX.